src/common/tensor_inspector.h

```diff
@@ -0,0 +1,763 @@
+    std::stringstream ss(str);
+    int i;
+    while (ss >> i) {
+      pos->push_back(i);
+      if (ss.peek() == ',') {
+        ss.ignore();
+      }
+    }
+    if (pos->size() > dimension) {
+      return false;
+    }
+    for (unsigned i = 0; i < pos->size(); ++i) {
+      if ((*pos)[i] > (tb_.shape_[i]-1)) {
+        return false;
+      }
+    }
+    return !pos->empty();
+  }
+
+  /*!
+   * \brief interactive print the tensor value
+   * \tparam DType the data type
+   * \param tag the name given to this call
+   */
+  template<typename DType>
+  void interactive_print_helper(std::string tag) {
+#if MXNET_USE_CUDA
+    if (tb_.dev_mask() == gpu::kDevMask) {
+      TensorInspector(test::CAccessAsCPU(ctx_, tb_, false)(), ctx_)
+          .interactive_print_helper<DType>(tag);
+      return;
+    }
+#endif  // MXNET_USE_CUDA
+    std::lock_guard<std::mutex> lock(InspectorManager::get()->mutex_);
+    InspectorManager::get()->interactive_print_tag_counter_[tag] += 1;
+    while (!InspectorManager::get()->interactive_print_skip_all_) {
+      std::cout << "----------Interactive Print----------" << std::endl;
+      if (tag != "") {
+        std::cout << "Tag: " << tag << "  Visit: " <<
+            InspectorManager::get()->interactive_print_tag_counter_[tag] <<  std::endl;
+      }
+      tensor_info_to_string<DType>(&std::cout);
+      std::cout << "Please specify the position, seperated by \",\"" << std::endl;
+      std::cout << "\"e\" for the entire tensor, " <<
+          "\"d\" to dump value to file, " <<
+          "\"b\" to break, " <<
+          "\"s\" to skip all: ";
+      std::string str;
+      std::cin >> str;
+      if (str == "b") {
+        break;
+      } else if (str == "e") {
+        to_string_helper<DType>(&std::cout);
+        continue;
+      } else if (str == "s") {
+        InspectorManager::get()->interactive_print_skip_all_ = true;
+        break;
+      } else if (str == "d") {
+        while (true) {
+          std::cout << "Please enter a tag: ";
+          std::cin >> str;
+          if (str.find(' ') != std::string::npos) {
+            std::cout << "Invalid input. ";
+            continue;
+          }
+          dump_to_file_helper<DType>(str);
+          break;
+        }
+        continue;
+      }
+      std::vector<int> pos;
+      if (parse_position(&pos, str)) {
+        std::vector<int> sub_shape;
+        size_t offset;
+        print_locator(pos, &sub_shape, &offset);
+        to_string_helper<DType>(&std::cout, sub_shape, offset);
+      } else {
+        std::cout << "invalid input" << std::endl;
+      }
+    }
+  }
+
+  /*!
+   * \brief build the lambda function, aka the checker, given its type
+   * \tparam DType the data type
+   * \param ct the type of the checker
+   */
+  template<typename DType>
+  std::function<bool(DType)> get_checker(CheckerType ct) {
+    switch (ct) {
+      case NegativeChecker:
+        return [] (DType x) {
+              return x < 0;
```

Comment:
I think the warnings are unavoidable if we want to have a templated function to return a bunch of value checkers (lambdas). C++ 17 has `constexpr` which can be used to our advantage to do compile type branching. I tried to simulate that using templates tricks (https://baptiste-wicht.com/posts/2015/07/simulate-static_if-with-c11c14.html) but it did not work on our compiler

Reply:
Using DEV = 1 will turn the warnings into errors and thus compilation will fail

Reply:
So I should fix those since developers will build with `DEV = 1`?

Reply:
since this is just a debugging tool i think it is fine.

Reply:
giving it a second look, stages like centos do build wiht make DEV=1 and they should fail in CI. I think its probably CI issue related to caching because of which they are not failing. Please check this otherwise CI may start failing after PR is merged.

Reply:
Sure